dm/pkg/checker/table_structure.go

```diff
@@ -296,168 +343,140 @@ func (c *ShardingTablesChecker) Check(ctx context.Context) *Result {
+				sourceID = checkItem.sourceID
+				p, err = dbutil.GetParserForDB(ctx, c.dbs[sourceID])
+				if err != nil {
+					c.reMu.Lock()
+					r.Extra = fmt.Sprintf("fail to get parser for sourceID %s on sharding %s", sourceID, c.targetTableID)
+					c.reMu.Unlock()
+					return err
+				}
+			}
+
+			statement, err := dbutil.GetCreateTableSQL(ctx, c.dbs[sourceID], table.Schema, table.Name)
+			if err != nil {
+				// continue if table was deleted when checking
+				if isMySQLError(err, mysql.ErrNoSuchTable) {
 					continue
 				}
+				return err
+			}
 
-				if r.Expression == column.PartitionID && r.TargetColumn == columnName {
-					hasMatchedRule = true
-					break
+			ctStmt, err := getCreateTableStmt(p, statement)
+			if err != nil {
+				return err
+			}
+
+			has := c.hasAutoIncrementKey(ctStmt)
+			if has {
+				c.reMu.Lock()
+				if r.State == StateSuccess {
+					r.State = StateWarning
+					r.Errors = append(r.Errors, NewError("sourceID %s table %v of sharding %s have auto-increment key, please make sure them don't conflict in target table!", sourceID, table, c.targetTableID))
+					r.Instruction = "If happen conflict, please handle it by yourself. You can refer to https://docs.pingcap.com/tidb-data-migration/stable/shard-merge-best-practices/#handle-conflicts-between-primary-keys-or-unique-indexes-across-multiple-sharded-tables"
+					r.Extra = AutoIncrementKeyChecking
 				}
+				c.reMu.Unlock()
 			}
 
-			if hasMatchedRule && !isBigInt {
+			checkErr := c.checkConsistency(c.firstCreateTableStmtNode, ctStmt, c.firstTable.String(), table.String(), c.firstSourceID, sourceID)
+			if checkErr != nil {
+				c.reMu.Lock()
 				r.State = StateFailure
-				r.Errors = append(r.Errors, NewError("instance %s table `%s`.`%s` of sharding %s have auto-increment key %s and column mapping, but type of %s should be bigint", instance, schema, table, c.name, columnName, columnName))
-				r.Instruction = "please set auto-increment key type to bigint"
-				r.Extra = AutoIncrementKeyChecking
-				return false
+				r.Errors = append(r.Errors, checkErr)
+				r.Extra = fmt.Sprintf("error on sharding %s", c.targetTableID)
+				r.Instruction = "please set same table structure for sharding tables"
+				c.reMu.Unlock()
+				return nil
 			}
 		}
-
-		if !hasMatchedRule {
-			r.State = StateFailure
-			r.Errors = append(r.Errors, NewError("instance %s table `%s`.`%s` of sharding %s have auto-increment key %s and column mapping, but type of %s should be bigint", instance, schema, table, c.name, columnName, columnName))
-			r.Instruction = "please handle it by yourself"
-			r.Extra = AutoIncrementKeyChecking
-			return false
-		}
 	}
-
-	return true
 }
 
-func (c *ShardingTablesChecker) findAutoIncrementKey(stmt *ast.CreateTableStmt, info *model.TableInfo) map[string]bool {
-	autoIncrementKeys := make(map[string]bool)
-	autoIncrementCols := make(map[string]bool)
-
+func (c *ShardingTablesChecker) hasAutoIncrementKey(stmt *ast.CreateTableStmt) bool {
```

Comment:
the previous code based on `mapping`(deprecated). If the task doesn't use `mapping`, the check will be failed. Users always disable it now.

So, we change it to report a warning if user set upstream auto-increment-key in sharding mode. And let user resolve it by themselves.